### PR TITLE
fixes #2639 After creating user with empty e-mail it becomes impossible ...

### PR DESCRIPTION
**Pull request by abenari**
...to add it.

### CODE REVIEW

**Comment by domcleal** [comment]
fine by me, @ohadlevy any reason you can think of to force presence of an e-mail address?

**Comment by ohadlevy** [comment]
there is a before_filter that force the user to add the mail, why didn't it jumped in?

**Comment by abenari** [comment]
The source of the bug is https://github.com/theforeman/foreman/commit/306ed25a518e027971739f2e9d182b1d6f72fb28 that fixes 
http://projects.theforeman.org/issues/1936
This is a minimal fix that let a logged-in  user without email, add his email and keep working.

**Comment by ohadlevy** [comment]
maybe i misunderstand what is the original bug? if there is no mail (OK to create without mail) user can still update it?

**Comment by abenari** [comment]
if there is no mail on create user, when that user logs-in we are preventing all actions, redirect the user to "my account" page where he is prompt to update the mail field. (All that logic is part of the fix by @rbirnie and is working)
The bug is that the user can't update the mail field, this patch fix it.
@domcleal is suggesting as mentioned on the irc, not to force mail in the first place, I tend to agree.
This patch however is a minimal patch and fixes the bug.

**Comment by abenari** [comment]
@domcleal replaced the solution for this bug, the source of the bug is commit 6a26fecd43e006007a4c92d49b9008be7d2e18c4 that validated the old object and prevented updating.

**Comment by domcleal** [comment]
@abenari there's a test failure here:

<pre>
expecting &lt;"edit"&gt; but rendering with &lt;""&gt; (test_0016_anonymous)
/var/lib/workspace/workspace/test_develop_pull_request/database/postgresql/puppet/3.0/ruby/1.8.7/test/functional/users_controller_test.rb:185 
</pre>

(http://ci.theforeman.org/job/test_develop_pull_request/2089/database=postgresql,puppet=3.0,ruby=1.8.7/testReport/junit/%28root%29/UsersControllerTest/test_0016_anonymous/)

I think you just need to remove the `unless` condition from the admin assignment, which is currently making the save appear successful and just ignoring the user if they try to remove the admin flag.  It doesn't work from the browser anyway as that sends `'0'` rather than `false`.

Could you add a test too please?

**Comment by ohadlevy** [comment]
[test]

File: test/functional/users_controller_test.rb [hunk 1]
```diff
@@ -185,4 +185,17 @@ def test_one #"should not remove the anonymous role" do
     assert_template :edit
   end
 
+  test "should be able to create user without mail and update the mail later" do
+     user = User.create :login => "mailess", :mail=> "", :auth_source => auth_sources(:one)
+
+     update_hash = {"user"=>{
+       "login"  => user.login,
+       "mail"  => "you@have.mail"},
+       "commit" => "Submit",
```
Comment (ohadlevy) [inline test/functional/users_controller_test.rb]: where does that comes from?

File: test/functional/users_controller_test.rb [hunk 2]
```diff
@@ -185,4 +185,17 @@ def test_one #"should not remove the anonymous role" do
...
+       "id"     => user.id}
```
Comment (ohadlevy) [inline test/functional/users_controller_test.rb:190]: not sure if you need the ID as well (as its not in the form?)

Comment (abenari) [inline test/functional/users_controller_test.rb:190]: both lines come from the test above.

Comment (abenari) [inline test/functional/users_controller_test.rb:190]: ID is needed.

**Comment by domcleal** [comment]
The test needed a small change in order to fail correctly, the user had to be an admin:

```
user.admin = true
user.save!(:validate => false)
```

I can fix during merge.

**Comment by domcleal** [comment]
Merged as 802ae177, thanks @abenari!

**Pull request closed without merging**